stylix: add `mkTarget` function (#1130)
nix-community/stylix: docs/src/modules.md

## Module template

Modules should be created using `mkTarget` whenever possible.

> When using `mkTarget`, request it as a non-destructured arg before the optional
> module args. See the below example.

Reviewer: Since you can't de-structure[^1] a function (unless it's a functor[^2]), it is redundant to mention de-structuring. Doing so in docs intended for new/external contributors will probably cause more confusion than clarity:

```suggestion
When using `mkTarget`, request it _before_ the optional module args.
See the below example.
```

[^1]: in the [functions section](https://nix.dev/manual/nix/2.28/language/syntax#functions) of the manual, they refer to de-structured args as a _"set pattern"_
[^2]: a functor is a callable attrset, see the [attrset section](https://nix.dev/manual/nix/2.28/language/syntax#attr-__functor) of the manual

> When impossible, modules should have an enable option created using `mkEnableTarget`.

Reviewer: ```suggestion
When `mkTarget` cannot be used, modules should declare an `enable` option using `mkEnableTarget`
and guard their config definitions behind the appropriate `mkIf` conditions.
```

This is similar to
[`mkEnableOption`](https://nix-community.github.io/docnix/reference/lib/options/lib-options-mkenableoption/)
from the standard library, however it integrates with
[`stylix.enable`](./options/nixos.md#stylixenable) and
[`stylix.autoEnable`](./options/nixos.md#stylixautoenable)
and generates more specific documentation.

General formats for modules are shown below.

Using `mkTarget`:

```nix
mkTarget:

> { config, lib, ... }: # this line for demonstration purposes only

Reviewer: ```suggestion
{ config, lib, ... }: # optional module args, after `mkTarget`

## Module template

Modules should be created using `mkTarget` whenever possible.

When using `mkTarget`, request it _before_ the optional module args.
See the below example.

When `mkTarget` cannot be used, modules should declare an `enable` option using `mkEnableTarget`
and guard their config definitions behind the appropriate `mkIf` conditions.
This is similar to
[`mkEnableOption`](https://nix-community.github.io/docnix/reference/lib/options/lib-options-mkenableoption/)
from the standard library, however it integrates with
[`stylix.enable`](./options/nixos.md#stylixenable) and
[`stylix.autoEnable`](./options/nixos.md#stylixautoenable)
and generates more specific documentation.

General formats for modules are shown below.

Using `mkTarget`:

```nix
mkTarget:

> { config, lib, ... }: # optional module args, after `mkTarget`

Reviewer: With the second approach, you would be able to implement additional restrictions on `config` if desired, e.g.

```nix
config = args.config // {
  stylix = throw "Do not access `config.stylix` in a target module!";
};
```

## Module template

Modules should be created using the `mkTarget` function whenever possible (see

> the `/stylix/mk-target.nix` documentation for more details):

Reviewer: ```suggestion
the in-source documentation for more details: [`/stylix/mk-target.nix`](https://github.com/danth/stylix/blob/-/stylix/mk-target.nix)):
```

```nix
{ config, lib, mkTarget ... }:
mkTarget {
  name = "example";
  humanName = "Example Target";

  configElements =
    { colors }:
    {
      programs.example.theme.background = colors.base00;
    };
}
```

> 

Reviewer: This is kinda an implementation detail, but it feels like it's work highlighting as it could trip someone up?

```suggestion

> [!IMPORTANT]
> The `mkTarget` argument is only available to modules imported by Stylix's [autoload system](https://github.com/danth/stylix/blob/-/stylix/autoload.nix),
> e.g., `modules/«target»/«platform».nix` modules.
>
> I.e., it is not available to normal modules imported via the `imports` list.

```

Reviewer: Is this the only place external contributors will commit code?

Even so, I still feel we should point out that it isn't available to normal modules.

It's possible a target module will import other modules in its directory, e.g. for refactoring complex logic. Such imported modules would not be able to specify a `mkTarget` module arg.

Also, contributors used to target modules may find it surprising they can't use `mkTarget` if they decide to submit a change elsewhere in Stylix, e.g. core features.

I agree we shouldn't get into the weeds or labour the point in these docs, but IMO it should be briefly highlighted because it is so unusual.

Reviewer 2: > It's possible a target module will import other modules in its directory, e.g. for refactoring complex logic. Such imported modules would not be able to specify a `mkTarget` module arg.
>
> Also, contributors used to target modules may find it surprising they can't use `mkTarget` if they decide to submit a change elsewhere in Stylix, e.g. core features.
>
> I agree we shouldn't get into the weeds or labour the point in these docs, but IMO it should be briefly highlighted because it is so unusual.

Makes sense. I would just wrap your suggestion at 80 characters:

```suggestion

> [!IMPORTANT]
> The `mkTarget` argument is only available to modules imported by Stylix's
> [autoload system](https://github.com/danth/stylix/blob/-/stylix/autoload.nix),
> e.g., `modules/«target»/«platform».nix` modules.
>
> I.e., it is not available to normal modules imported via the `imports` list.

```

When the `mkTarget` function cannot be used, modules must manually replicate its

> safeguarding functionality:

Reviewer: ```suggestion
safeguarding checks:
```

```suggestion
safeguarding conditions:
```

```suggestion
safeguarding behaviour:
```

```nix
{ config, lib, ... }:
{

>   options.stylix.targets.example.enable =
>     config.lib.stylix.mkEnableTarget "Example Target" true;
> 
>   config =
>     lib.mkIf (config.stylix.enable && config.stylix.targets.example.enable)
>       {
>         programs.example.theme.background = config.lib.stylix.colors.base00;
>       };
> }

Reviewer: IMO it's clearer with the «unicode» arrows indicating placeholder values, I don't see any reason to change this example in this PR...

```suggestion
  options.stylix.targets.«name».enable =
    config.lib.stylix.mkEnableTarget "«human readable name»" true;

  config =
    lib.mkIf (config.stylix.enable && config.stylix.targets.«name».enable)
      {
        programs.«name».backgroundColor = config.lib.stylix.colors.base00;
      };
}
```

If anything, maybe the `mkTarget` example above should use the same placeholder pattern?

## Module template

> All modules should be created using `mkTarget`.

Reviewer 3: mkTarget doesn't cover some edge cases e.g. the firefox module. We might want to leave a note about that.

A general format for modules is shown below.

```nix
mkTarget:

> { config, lib, ... }:

Reviewer 3: ```suggestion
```

Reviewer: I agree _this_ example doesn't need module args. But demonstrating they can be used is useful.

Maybe another example could demonstrate them by using `lib` for declaring an option or something?

Author: I added a note about this. I would like to keep it there, even though it is unnecessary, for the purposes of demonstration.

mkTarget
{
  name = "«target attribute name»";
  humanName = "«human readable name»";

  configElements = [
    { colors }:
      {

>         programs.«name».backgroundColor = config.lib.stylix.colors.base00;

Reviewer 3: ```suggestion
        programs.«name».backgroundColor = colors.base00;
```

>   configElements = [
>     { colors }:
>       {
>         programs.«name».backgroundColor = config.lib.stylix.colors.base00;
>       };
>   ];

Reviewer: This has a syntax error, list element should be wrapped with `(`parens`)` because it contains whitespace.

Alternatively, you could use the new non-list style as it is a single element:

```suggestion
  configElements =
    { colors }:
    {
      programs.«name».backgroundColor = config.lib.stylix.colors.base00;
    };
```

}
```

### `mkTarget`

The `mkTarget` function is a special function which automatically generates
options and guards configuration with`stylix.${attr} != null` on configuration
to cut down on boilerplate.

Arguments without defaults must be set explicitly.

> | Argument | Type | Default | Description | Example |

Reviewer 3: please format this table to make it more readable when viewed as plain text

Reviewer: pros & cons to having the column dividers aligned:
- future changes that edit one line may need to reformat the entire table
- the table may end up _very_ wide
- the table is easier to read
- easier to tell which column a specific field is in

Even without strictly aligning the columns, it should be possible to format slightly more readably.

|----------|------|---------|-------------|---------|
|`name`|String||Used when generating options; `stylix.targets.${name}`|"firefox"|
|`humanName`|String||Used when generating enable option; `stylix.targets.${name}.enable` has a description of "Whether to enable theming for ${humanName}."|"Firefox"|

> |`autoEnableConditions`|Boolean|true|Used for determining when an attribute called `autoEnable` should be set to `false` if one of the following applies:<ul><li>The module requires further manual setup to work correctly.</li><li>There is no reliable way to detect whether the target is installed, *and* enabling it unconditionally would cause problems.</li></ul>|`pkgs.stdenv.hostPlatform.isLinux`|

Reviewer: Note: if we're using/recommending dynamic values for `autoEnable` we need to think about how that should be documented. See some of the FIXMEs that came up in #1212, for example.

This is already an issue with `mkEnableTarget`, but previously the recommendation was simply `true` or `false`.

Eventually, we'll probably need some kinda `autoEnableText` that can be incorporated into the relevant options' defaultText & example literalExpressions (or literalMDs).

Maybe for now we should avoid suggesting dynamic values?

> |`extraOptions`|Attrset of options|{ }|Attributes will be mapped to `stylix.targets.${name}.${option}`|`{ colorTheme.enable = lib.mkEnableOption "color theme"; }`|

Reviewer: ```suggestion
|`extraOptions`|Attrset of option declarations|{ }|Attributes will be mapped to `stylix.targets.${name}.${option}`|`{ colorTheme.enable = lib.mkEnableOption "color theme"; }`|
```

When mentioning options its usually a good idea to clarify whether we mean declarations (`options.foo = mkOption {}`) or definitions (`config.foo = "bar"`).

Unrelated: is it important to name the argument `extraOptions` instead of simply `options`? I guess it makes clear this is a custom abstraction. Maybe `targetOptions` is a better name, although that should be clear from the fact we're calling `mkTarget` :thinking: ?

Author: The reasoning behind the name was to not mislead contributors into thinking they have to make the `enable` option themselves. If you think this would not be a problem, I will change the name to `options`.

Reviewer: > not mislead contributors into thinking they have to make the enable option themselves.

Makes sense. In that case, I agree with avoiding `options`.

Something like `extraOptions`, `targetOptions`, etc is clearer 👍

> |`configElements`|List of or single (attrset or function returning attrset)|{ }|Function(s) should take the Stylix attributes that it needs (e.g. `colors`, `image`, `fonts`) and `cfg` if necessary. This `cfg` is an alias for `config.stylix.targets.${name}`. Function(s) should return an attribute set to be merged into `config`. If any of the Stylix attributes needed by a/the function are `null`, the configuration will not be merged.|*See above template*|

Reviewer 3: should also note somewhere that `config.stylix.*` should never be accessed inside of configElements through `config`

> |`configElements`|List of function returning attrset|{ }|Each function should take the Stylix attributes that it needs (e.g. `colors`, `image`, `fonts`) and `cfg` if necessary. This `cfg` is an alias for `config.stylix.targets.${name}`. Each function should return an attribute set to be merged into `config`. If any of the Stylix attributes needed by a function are `null`, the configuration will not be merged.|*See above template*|

Reviewer: Nice to haves:

It being a list could be optional; we could normalise internally with `lib.toList`

Elememts being a function could be optional; we could check using `isFunction` and/or `isAttrs`.

> |`generalConfig`|Null or function returning attrset|null|Same format as functions under `configElements`. Merged regardless of whether requested Stylix options are `null`. Avoid use when possible.|*See above template*|

Reviewer: Q: in what scenario(s) would requested attrs be null? As an outsider this isn't clear to me.

Is this related to optional third-party modules, e.g., nixvim's home-manager module?

Not easy to come up with a concise description, but maybe the intention could be made clearer, or a link to more in-depth docs could be added?

Author: When Stylix options are optional, having them set as `null` is the default and equivalent to them being unset. When they are unset, all theming relating to them is disabled. Currently, there are only a few optional options, but our goal is to make all options optional. With this function, much of the legwork and repeated guards are handled for us.

See https://github.com/danth/stylix/discussions/1009#discussion-8091577 under **impetus** for more in-depth explanation. I have added two words to the description which hopefully give some clarification on the intention, but I will add more if you think it is necessary.

> |`generalConfig`|Null or function returning attrset|null|Same function format as in `configElements`, except the result of this function will always be merged with `config`. Avoid use when possible.|*See above template*|

Reviewer: Same: could be null or attrs or function.

Also: I wonder if the wording could be simplified? Maybe something along the lines of "unconditional config" that is "always defined" "regardless of enable options" 🤔 ?

Author: Wording was unclear, `generalConfig` is only different from `configElements` in that it does not have its attribute requests checked if they are null. Because of that, there is no reason for it to be able to take attrs as it would be duplicated behavior from `configElements`
